workshops/views.py

```diff
@@ -145,51 +148,201 @@ def person_bulk_add(request):
         form = PersonBulkAddForm(request.POST, request.FILES)
         if form.is_valid():
             try:
-                persons_tasks = _upload_person_task_csv(request, request.FILES['file'])
+                persons_tasks = _upload_person_task_csv(request,
+                                                        request.FILES['file'])
             except csv.Error as e:
-                messages.add_message(request, messages.ERROR, "Error processing uploaded .CSV file: {}".format(e))
+                messages.add_message(request, messages.ERROR,
+                                     "Error processing uploaded .CSV file: {}"
+                                     .format(e))
             else:
-                for entry in persons_tasks:
-                    entry['person'].save()
-                    if entry['task']:
-                        entry['task'].person = entry['person'] # Because Django's ORM doesn't do this automatically.
-                        entry['task'].save()
-                context = {'title' : 'Process CSV File',
-                           'form': form, 
-                           'persons_tasks': persons_tasks}
-                return render(request, 'workshops/person_bulk_add_results.html', context)
+                # instead of insta-saving, put everything into session
+                # then redirect to confirmation page which in turn saves the
+                # data
+
+                request.session['bulk-add-people'] = persons_tasks
+
+                return redirect('person_bulk_add_confirmation')
+
     else:
         form = PersonBulkAddForm()
 
-    context = {'title' : 'Bulk Add People',
+    context = {'title': 'Bulk Add People',
                'form': form}
     return render(request, 'workshops/person_bulk_add_form.html', context)
 
 
+def person_bulk_add_confirmation(request):
+    """
+    This view allows for manipulating and saving session-stored upload data.
+    """
+    persons_tasks = request.session['bulk-add-people']
+
+    # if the session is empty, don't bother
+    if not persons_tasks:
+        raise Http404
```

Comment:
I don't think the 404 would be helpful and it's not really semantically correct. I think it'd be better to add a warning message about not having the csv data from the previous operation and to try again, then redirect to the bulk upload page.